app/src/main/java/duke/Duke.java

```diff
@@ -0,0 +1,138 @@
+package duke;
+
+import java.time.format.DateTimeParseException;
+import java.util.NoSuchElementException;
+
+public class Duke {
+    private static boolean isAlive;
+    private static Ui ui;
+    private static Storage storage;
+    private static Parser parser;
+    private static TaskList tasks;
+
+    private static int checkTask(String id) {
+        try {
+            int task = Integer.parseInt(id);
+            if (task > tasks.size() || task <= 0) {
+                return -1;
+            }
+            return task;
+        } catch (NumberFormatException e) {
+            return -2;
+        }
+    }
+
+    private static String handle(String command, String params) {
+        switch (command) {
+        case "list":
+            StringBuilder out = new StringBuilder();
+            for (int i = 0; i < tasks.size(); ++i) {
+                if (i != 0) {
+                    out.append("\n");
+                }
+                out.append(i + 1).append(". ").append(tasks.getTask(i));
+            }
+            if (out.toString().equals("")) {
+                return "No tasks left!";
+            }
+            return out.toString();
+        case "find":
+            StringBuilder searchResults = new StringBuilder();
+            for (int i = 0; i < tasks.size(); ++i) {
+                if (!tasks.getTask(i).toString().contains(params)) {
+                    continue;
+                }
+                if (searchResults.length() != 0) {
+                    searchResults.append("\n");
+                }
+                searchResults.append(i + 1).append(". ").append(tasks.getTask(i));
+            }
+            if (searchResults.toString().equals("")) {
+                return "No search results!";
+            }
+            return searchResults.toString();
+        case "mark":
+            int markedTask = checkTask(params);
+            if (markedTask < 0) {
+                return "Invalid task number!";
```

Comment:
Another alternative is to create Exception subclass to handle invalid input and also store the error message. You could also make different subclass for different cases such as invalid date input, unknown command.